README.md
# VintageNetWireguard

**Experimental**

An attempt to support Wireguard VPN peer connections on Nerves devices. See
https://wireguard.com for more info

## Configuration
<!--- DOC !--->
Wireguard needs to configure an interface and peer connections for that
interface. Below is a list of the expected configuration parameters
which are referenced from [`wg(8)`](https://git.zx2c4.com/wireguard-tools/about/src/man/wg.8)
and [`wg-quick(8)`](https://git.zx2c4.com/wireguard-tools/about/src/man/wg-quick.8):

### Interface

| Key | `wg` name | Required? | Description |
| --- | --- | :---: | --- |
| `:private_key` | `PrivateKey` | X | base64 private key for the interface registered with the server |
| `:addresses` | `Address` | X | list of IP addresses for the connection to use (CIDR supported) |
| `:listen_port` | `ListenPort` | | port for the connection. Randomly assigned if empty or `0` |
| `:fwmark` | `FwMark` | | 32-bit fwmark for outgoing packets |
| `:dns` | `DNS` | | list of DNS IP's |
| `:peers` | `[PEER]` | | list of peer configs (see below) |

### Peer

| Key | `wg` name | Required? | Description |
| --- | --- | :---: | --- |
| `:public_key` | `PublicKey` | X | base64 public key |
| `:endpoint` | `Endpoint` | X | endpoint to the wireguard server which the peer attempts to connect |
| `:allowed_ips` | `AllowedIps` | X | list of IP addresses for allowed incoming packets and outgoing packets directed to. Defaults to `["0.0.0.0/0", "::0"]` |
| `:persistent_keepalive` | `PersistentKeepalive` | | optional integer seconds for sending an authenticated packet as a keepalive |

## Using Wireguard Config Files

Wireguard commonly uses `*.conf` configuration files to simplify the setup
process and `VintageNetWireguard` provides a helper function to parse those
config files into the expected format:

```elixir
iex)> config = VintageNetWireguard.ConfigFile.parse("/path/to/wg0.conf")
iex)> VintageNet.configure("wg0", config)
```
<!--- DOC !--->
## Goals/Ideas

- [X] Setup `wg*` network interfaces
- [ ] Notes/cookbook for setting up Wireguard server
  - [ ] fly.io
  - [ ] Another service?
- [X] Parse wireguard peer configs
- [ ] Potentially support authentication via other routes (i.e. NervesKey)
- [ ] Mechanism for registering a new peer with remote server
- [ ] Prevent storing private keys on disc


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.2.0] - 2023-12-26

### Added

* Scope logger messages to module
* Include stderr with `wg` command output
* Support OTP >= 25

### Fixed

* Ensure peer arg ordering is correct
* Change `allowed-ips` to expected comma-separated list
* Fix warnings for Elixir >= 1.15

### Updated

* [wireguard-tools@13f4ac4](https://git.zx2c4.com/wireguard-tools/commit/?id=13f4ac4cb74b5a833fa7f825ba785b1e5774e84f)

[v0.2.0]: https://github.com/nerves-networking/vintage_net_wireguard/releases/tag/v0.2.0

## [v0.1.0] - 2022-02-22

Initial release with support for starting a Wireguard client connection

See the [README.md](README.md) for more information on configuration

### Added

* `VintageNetWireguard.ConfigFile.parse/1` for parsing a Wireguard `*.conf` file
  into the expected format for configuring with VintageNet
* Compilation of [`wg(8)`](https://git.zx2c4.com/wireguard-tools/about/src/man/wg.8)
  utility

[v0.1.0]: https://github.com/nerves-networking/vintage_net_wireguard/releases/tag/v1.0.0


test/vintage_net_wireguard_test.exs
defmodule VintageNetWireguardTest do
  use ExUnit.Case
  doctest VintageNetWireguard

  @peer %{endpoint: "somewhere", public_key: "pubkey"}

  @config %{
    private_key: "key",
    addresses: [],
    peers: [@peer],
    fwmark: 1234,
    listen_port: 5678,
    type: VintageNetWireguard
  }

  describe "normalize/1" do
    test "requires :private_key" do
      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(Map.delete(@config, :private_key))
      end

      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(%{@config | private_key: ""})
      end
    end

    test "requires :addresses" do
      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(Map.delete(@config, :addresses))
      end

      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(%{@config | addresses: nil})
      end

      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(%{@config | addresses: "10.0.0.1"})
      end
    end

    test "requires :peers" do
      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(Map.delete(@config, :peers))
      end

      assert_raise ArgumentError, fn -> VintageNetWireguard.normalize(%{@config | peers: nil}) end

      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(%{@config | peers: "10.0.0.1"})
      end
    end

    test "peer requires endpoint" do
      assert_raise ArgumentError, fn ->
        bad = Map.delete(@peer, :endpoint)
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end

      assert_raise ArgumentError, fn ->
        bad = %{@peer | endpoint: ""}
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end

      assert_raise ArgumentError, fn ->
        bad = %{@peer | endpoint: nil}
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end
    end

    test "peer requires public_key" do
      assert_raise ArgumentError, fn ->
        bad = Map.delete(@peer, :public_key)
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end

      assert_raise ArgumentError, fn ->
        bad = %{@peer | public_key: ""}
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end

      assert_raise ArgumentError, fn ->
        bad = %{@peer | public_key: nil}
        VintageNetWireguard.normalize(%{@config | peers: [bad]})
      end
    end

    test "valid addresses" do
      ipv4 = {{172, 31, 78, 149}, 32}
      ipv6 = {{65152, 0, 0, 0, 50397, 60671, 65200, 26324}, 128}

      addrs = [
        "172.31.78.149",
        "172.31.78.149/32",
        {172, 31, 78, 149},
        ipv4,
        "fe80::c4dd:ecff:feb0:66d4/128",
        "fe80::c4dd:ecff:feb0:66d4",
        {65152, 0, 0, 0, 50397, 60671, 65200, 26324},
        ipv6
      ]

      expected = [
        ipv4,
        ipv4,
        ipv4,
        ipv4,
        ipv6,
        ipv6,
        ipv6,
        ipv6
      ]

      assert %{addresses: ^expected} =
               VintageNetWireguard.normalize(%{@config | addresses: addrs})
    end

    test "valid fwmark" do
      fwmark = @config.fwmark
      assert %{fwmark: ^fwmark} = VintageNetWireguard.normalize(@config)
    end

    test "ignores invalid fwmark" do
      invalid = [nil, [], "foo", "1234", "0x1234"]

      for bad <- invalid do
        config = VintageNetWireguard.normalize(%{@config | fwmark: bad})
        refute Map.has_key?(config, :fwmark)
      end
    end

    test "valid listen_port" do
      listen_port = @config.listen_port
      assert %{listen_port: ^listen_port} = VintageNetWireguard.normalize(@config)
    end

    test "ignores invalid listen_port" do
      invalid = [nil, [], "foo", "1234", "0x1234"]

      for bad <- invalid do
        config = VintageNetWireguard.normalize(%{@config | listen_port: bad})
        refute Map.has_key?(config, :listen_port)
      end
    end

    test "ignores invalid :dns field" do
      config = VintageNetWireguard.normalize(Map.put(@config, :dns, "foo"))
      config2 = VintageNetWireguard.normalize(Map.put(@config, :dns, nil))

      refute Map.has_key?(config, :dns)
      refute Map.has_key?(config2, :dns)
    end

    test "fails with invalid DNS entries" do
      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(Map.put(@config, :dns, ["foo"]))
      end

      assert_raise ArgumentError, fn ->
        VintageNetWireguard.normalize(Map.put(@config, :dns, [nil]))
      end
    end

    test "valid :dns" do
      ipv4 = {172, 31, 78, 149}
      ipv6 = {65152, 0, 0, 0, 50397, 60671, 65200, 26324}

      dns = [
        "172.31.78.149",
        ipv4,
        "fe80::c4dd:ecff:feb0:66d4",
        ipv6
      ]

      expected = [
        ipv4,
        ipv4,
        ipv6,
        ipv6
      ]

      assert %{dns: ^expected} = VintageNetWireguard.normalize(Map.put(@config, :dns, dns))
    end

    test "valid peer preshared key" do
      p = Map.put(@peer, :preshared_key, "key")

      assert %{peers: [%{preshared_key: "key"}]} =
               VintageNetWireguard.normalize(%{@config | peers: [p]})
    end

    test "ignores invalid peer preshared key" do
      invalid = [nil, [], 123_445]

      for bad <- invalid do
        p = Map.put(@peer, :preshared_key, bad)
        %{peers: [updated]} = VintageNetWireguard.normalize(%{@config | peers: [p]})
        refute Map.has_key?(updated, :preshared_key)
      end
    end

    test "valid peer persistent_keepalive" do
      p = Map.put(@peer, :persistent_keepalive, 1234)

      assert %{peers: [%{persistent_keepalive: 1234}]} =
               VintageNetWireguard.normalize(%{@config | peers: [p]})
    end

    test "ignores invalid peer persistent_keepalive" do
      invalid = [nil, [], "123445", :bad]

      for bad <- invalid do
        p = Map.put(@peer, :persistent_keepalive, bad)
        %{peers: [updated]} = VintageNetWireguard.normalize(%{@config | peers: [p]})
        refute Map.has_key?(updated, :persistent_keepalive)
      end
    end
  end
end


mix.exs
defmodule VintageNetWireguard.MixProject do
  use Mix.Project

  @version "0.2.0"
  @source_url "https://github.com/nerves-networking/vintage_net_wireguard"

  def project() do
    [
      app: :vintage_net_wireguard,
      version: @version,
      elixir: "~> 1.11",
      compilers: [:elixir_make | Mix.compilers()],
      make_targets: ["all"],
      make_clean: ["mix_clean"],
      make_error_message: "",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      dialyzer: dialyzer(),
      docs: docs(),
      package: package(),
      preferred_cli_env: [
        docs: :docs,
        "hex.build": :docs,
        "hex.publish": :docs
      ]
    ]
  end

  def application() do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps() do
    [
      {:vintage_net, "~> 0.11"},
      {:temp, "~> 0.4"},
      {:net_address, "~> 0.3"},
      {:elixir_make, "~> 0.6", runtime: false},
      {:dialyxir, "~> 1.1", only: [:dev, :test], runtime: false},
      {:credo, "~> 1.6", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.28", only: :docs, runtime: false}
    ]
  end

  defp description() do
    "Wireguard VPN support for VintageNet"
  end

  defp dialyzer() do
    [
      flags: [:missing_return, :extra_return, :unmatched_returns, :error_handling, :underspecs]
    ]
  end

  defp docs() do
    [
      extras: ["README.md", "CHANGELOG.md"],
      main: "readme",
      source_ref: "v#{@version}",
      source_url: @source_url,
      skip_undefined_reference_warnings_on: ["CHANGELOG.md"]
    ]
  end

  defp package() do
    %{
      files: [
        "lib",
        "mix.exs",
        "Makefile",
        "README.md",
        "LICENSE",
        "CHANGELOG.md"
      ],
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => @source_url}
    }
  end
end


lib/vintage_net_wireguard/config_file.ex
defmodule VintageNetWireguard.ConfigFile do
  @moduledoc """
  Helpers for dealing with a wireguard configuration file
  """

  @doc """
  Parse a Wireguard config file into a map
  """
  @spec parse(Path.t()) :: map() | {:error, File.posix()}
  def parse(path) do
    case File.read(path) do
      {:ok, file} ->
        file
        |> String.split("\n")
        |> parse_lines(%{type: VintageNetWireguard})

      err ->
        err
    end
  end

  defp parse_lines([], acc), do: acc

  defp parse_lines(["[Peer]" <> _ | rest], acc) do
    peers = acc[:peers] || []
    {peer, rem} = parse_peer_lines(rest, %{})
    parse_lines(rem, Map.put(acc, :peers, [peer | peers]))
  end

  defp parse_lines(["PrivateKey" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :private_key, parse_val(bin)))
  end

  defp parse_lines(["ListenPort" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> String.to_integer()

    parse_lines(rest, Map.put(acc, :listen_port, val))
  end

  defp parse_lines(["FwMark" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :fw_mark, parse_val(bin)))
  end

  defp parse_lines(["Address" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> comma_separated()

    parse_lines(rest, Map.put(acc, :addresses, val))
  end

  defp parse_lines(["DNS" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> comma_separated()

    parse_lines(rest, Map.put(acc, :dns, val))
  end

  defp parse_lines(["MTU" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> String.to_integer()

    parse_lines(rest, Map.put(acc, :mtu, val))
  end

  defp parse_lines(["Table" <> bin | rest], acc) do
    val = if parse_val(bin) == "off", do: :off, else: :auto
    parse_lines(rest, Map.put(acc, :table, val))
  end

  defp parse_lines(["PreUp" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :pre_up, parse_val(bin)))
  end

  defp parse_lines(["PostUp" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :post_up, parse_val(bin)))
  end

  defp parse_lines(["PreDown" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :pre_down, parse_val(bin)))
  end

  defp parse_lines(["PostDown" <> bin | rest], acc) do
    parse_lines(rest, Map.put(acc, :post_down, parse_val(bin)))
  end

  defp parse_lines([_ | rest], acc), do: parse_lines(rest, acc)

  defp parse_peer_lines([], acc), do: {acc, []}

  defp parse_peer_lines(["[Peer]" <> _ | _] = rest, acc) do
    {acc, rest}
  end

  defp parse_peer_lines(["AllowedIPs" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> comma_separated()

    parse_peer_lines(rest, Map.put(acc, :allowed_ips, val))
  end

  defp parse_peer_lines(["PublicKey" <> bin | rest], acc) do
    parse_peer_lines(rest, Map.put(acc, :public_key, parse_val(bin)))
  end

  defp parse_peer_lines(["PresharedKey" <> bin | rest], acc) do
    parse_peer_lines(rest, Map.put(acc, :preshared_key, parse_val(bin)))
  end

  defp parse_peer_lines(["Endpoint" <> bin | rest], acc) do
    parse_peer_lines(rest, Map.put(acc, :endpoint, parse_val(bin)))
  end

  defp parse_peer_lines(["PersistentKeepalive" <> bin | rest], acc) do
    val =
      parse_val(bin)
      |> String.to_integer()

    parse_peer_lines(rest, Map.put(acc, :persistent_keepalive, val))
  end

  defp parse_peer_lines([_ | rest], acc), do: parse_peer_lines(rest, acc)

  defp parse_val(line) do
    Regex.run(~r/^\s?=\s?(.*)/, line, capture: :all_but_first)
    |> hd()
  end

  defp comma_separated(val), do: String.split(val, ~r/\s?,\s?/)
end


lib/vintage_net_wireguard.ex
defmodule VintageNetWireguard do
  @moduledoc File.read!("README.md")
             |> String.split("<!--- DOC !--->")
             |> Enum.at(1)

  @behaviour VintageNet.Technology

  import IP, only: [is_ip: 1, is_ipv4: 1, is_ipv6: 1]

  alias VintageNet.Interface.RawConfig

  require Logger

  defguard is_non_empty_string(str) when is_binary(str) and str != ""

  @required_config [
    private_key: :string,
    addresses: :list,
    peers: :list
  ]

  @required_peer [
    endpoint: :string,
    public_key: :string
  ]

  @default_allowed_ips [
    # 0.0.0.0/0
    {{0, 0, 0, 0}, 0},
    # ::/0
    {{0, 0, 0, 0, 0, 0, 0, 0}, 0}
  ]

  @impl VintageNet.Technology
  def normalize(%{type: __MODULE__} = config) do
    _ = ensure_required!(config, @required_config)

    config
    |> Map.put(:addresses, Enum.map(config.addresses, &normalize_address/1))
    |> normalize_fwmark()
    |> normalize_listen_port()
    |> normalize_dns()
    |> Map.put(:peers, Enum.map(config.peers, &normalize_peer/1))
  end

  defp ensure_required!(config, schema) do
    for {key, type} <- schema do
      case {config[key], type} do
        {v, :string} when is_non_empty_string(v) -> :ok
        {v, :list} when is_list(v) -> :ok
        _ -> raise ArgumentError, ":#{key} is required"
      end
    end
  end

  defp normalize_address(addr) when is_binary(addr) do
    result =
      if addr =~ ~r/\/\d+$/ do
        IP.Subnet.config_from_string(addr)
      else
        IP.from_string(addr)
      end

    case result do
      {:ok, ip, %{bit_length: len}} ->
        {ip, len}

      {:ok, ip} ->
        normalize_address(ip)

      {:error, err} ->
        raise ArgumentError, "Unable to parse address #{addr}: #{inspect(err)}"
    end
  end

  defp normalize_address({ip, prefix_length} = addr) when is_ip(ip) and is_integer(prefix_length),
    do: addr

  defp normalize_address(ip) when is_ipv6(ip), do: {ip, 128}
  defp normalize_address(ip) when is_ipv4(ip), do: {ip, 32}

  defp normalize_address(addr) do
    raise ArgumentError, "invalid address #{inspect(addr)}"
  end

  defp normalize_fwmark(%{fwmark: fwmark} = config) when is_integer(fwmark), do: config

  defp normalize_fwmark(%{fwmark: fwmark} = config) do
    Logger.warning("[VintageNetWireguard] Ignoring invalid FwMark: #{inspect(fwmark)}")
    Map.delete(config, :fwmark)
  end

  defp normalize_fwmark(config), do: config

  defp normalize_listen_port(%{listen_port: port} = config) when is_integer(port), do: config

  defp normalize_listen_port(%{listen_port: port} = config) do
    Logger.warning("[VintageNetWireguard] Ignoring invalid ListenPort: #{inspect(port)}")
    Map.delete(config, :listen_port)
  end

  defp normalize_listen_port(config), do: config

  defp normalize_dns(%{dns: dns} = config) when is_list(dns) do
    dns = for d <- dns, {ip, _prefix_len} = normalize_address(d), do: ip
    %{config | dns: dns}
  end

  defp normalize_dns(%{dns: dns} = config) do
    Logger.warning("[VintageNetWireguard] Ignoring invalid DNS: #{inspect(dns)}")
    Map.delete(config, :dns)
  end

  defp normalize_dns(config), do: config

  defp normalize_peer(peer) do
    _ = ensure_required!(peer, @required_peer)
    aips = peer[:allowed_ips] || @default_allowed_ips

    Map.put(peer, :allowed_ips, Enum.map(aips, &normalize_address/1))
    |> normalize_keepalive()
    |> normalize_preshared_key()
  end

  defp normalize_keepalive(%{persistent_keepalive: pk} = peer) when pk in 0..65535, do: peer

  defp normalize_keepalive(%{persistent_keepalive: pk} = peer) do
    Logger.warning(
      "[VintageNetWireguard] Ignoring invalid peer PersistentKeepalive: #{inspect(pk)}"
    )

    Map.delete(peer, :persistent_keepalive)
  end

  defp normalize_keepalive(peer), do: peer

  defp normalize_preshared_key(%{preshared_key: psk} = peer) when is_non_empty_string(psk),
    do: peer

  defp normalize_preshared_key(%{preshared_key: psk} = peer) do
    Logger.warning("[VintageNetWireguard] Ignoring invalid peer PresharedKey: #{inspect(psk)}")
    Map.delete(peer, :preshared_key)
  end

  defp normalize_preshared_key(peer), do: peer

  @impl VintageNet.Technology
  def to_raw_config(ifname, %{type: __MODULE__} = config, _opts) do
    normalized = normalize(config)

    %RawConfig{
      ifname: ifname,
      type: __MODULE__,
      source_config: normalized,
      child_specs: [{VintageNet.Connectivity.LANChecker, ifname}],
      required_ifnames: [],
      up_cmds: up_cmds(ifname, normalized),
      up_cmd_millis: 20_000,
      down_cmds: [
        {:fun, VintageNet.RouteManager, :clear_route, [ifname]},
        {:fun, VintageNet.NameResolver, :clear, [ifname]},
        {:run_ignore_errors, "ip", ["addr", "flush", "dev", ifname, "label", ifname]},
        {:run, "ip", ["link", "set", ifname, "down"]}
      ]
    }
  end

  @impl VintageNet.Technology
  def check_system(_opts), do: {:error, "unimplemented"}

  @impl VintageNet.Technology
  def ioctl(_ifname, _cmd, _args), do: {:error, :unsupported}

  @doc """
  Path to wg executable
  """
  @spec wg() :: Path.t()
  def wg(), do: Path.join(:code.priv_dir(:vintage_net_wireguard), "wg")

  defp up_cmds(ifname, config) do
    add_addresses =
      for addr <- config.addresses do
        {:run, "ip", ["addr", "add", addr_subnet(addr), "dev", ifname, "label", ifname]}
      end

    [
      maybe_add_interface(ifname),
      {:fun, fn -> set_wg_interface(ifname, config) end},
      {:run_ignore_errors, "ip", ["addr", "flush", "dev", ifname, "label", ifname]},
      add_addresses,
      {:run, "ip", ["link", "set", ifname, "up"]},
      Enum.map(config.peers, &peer_commands(ifname, &1)),
      maybe_add_dns(ifname, config),
      # Make MTU configurable?
      {:run, "ip", ["link", "set", "mtu", "1420", "up", "dev", ifname]}
    ]
    |> List.flatten()
  end

  defp maybe_add_interface(ifname) do
    case System.cmd("ip", ["link", "show", ifname]) do
      {_, 0} -> []
      _ -> {:run_ignore_errors, "ip", ["link", "add", ifname, "type", "wireguard"]}
    end
  end

  # TODO: ensure cleanup and fix this
  # Temp specs say this should return :ok, but it actually returns
  # a list paths. Need to suppress dialyzer until Temp is fixed
  # or alternate cleanup is implemented
  @dialyzer {:nowarn_function, set_wg_interface: 2}

  defp set_wg_interface(ifname, config) do
    with {:ok, tracker} <- Temp.track(),
         if_args = Enum.reduce(config, [], &add_if_arg/2),
         {_, 0} <- System.cmd(wg(), ["set", ifname | if_args], stderr_to_stdout: true),
         [_ | _] <- Temp.cleanup(tracker) do
      :ok
    else
      err ->
        Temp.cleanup()
        {:error, "[VintageNetWireguard] Failed to set interface - #{inspect(err)}"}
    end
  end

  defp add_if_arg({:private_key, key}, acc) do
    {:ok, path} = Temp.open(%{}, &IO.write(&1, key))
    ["private-key", path | acc]
  end

  defp add_if_arg({:fwmark, v}, acc), do: ["fwmark", "#{v}" | acc]
  defp add_if_arg({:listen_port, v}, acc), do: ["listen-port", "#{v}" | acc]
  defp add_if_arg(_, acc), do: acc

  defp peer_commands(ifname, peer) do
    [
      {:fun, fn -> set_peer(ifname, peer) end},
      # TODO: Maybe use VintageNet.RouteManager when IPv6 is supported?
      # {:fun, VintageNet.RouteManager, :set_route, [ifname, peer.allowed_ips, peer.endpoint]}
      Enum.map(peer.allowed_ips, &route_ip(ifname, &1))
    ]
  end

  defp set_peer(ifname, peer) do
    with {:ok, tracker} <- Temp.track(),
         # peer arg needs to be first
         peer_args = ["peer", peer.public_key | Enum.reduce(peer, [], &add_peer_arg/2)],
         {_, 0} <- System.cmd(wg(), ["set", ifname | peer_args], stderr_to_stdout: true) do
      _ = Temp.cleanup(tracker)
      :ok
    else
      {err, s} ->
        Logger.error("""
        [VintageNetWireguard] Nonzero exit setting peer: #{s}

        #{inspect(err)}
        """)

        _ = Temp.cleanup()

        {:error, :non_zero_exit}
    end
  end

  defp add_peer_arg({:allowed_ips, v}, acc) do
    addrs = Enum.map_join(v, ",", &addr_subnet/1)
    ["allowed-ips", addrs | acc]
  end

  defp add_peer_arg({:endpoint, v}, acc), do: ["endpoint", v | acc]
  defp add_peer_arg({:persistent_keepalive, v}, acc), do: ["persistent-keepalive", "#{v}" | acc]

  defp add_peer_arg({:preshared_key, v}, acc) do
    {:ok, path} = Temp.open(%{}, &IO.write(&1, v))
    ["preshared-key", path | acc]
  end

  defp add_peer_arg(_, acc), do: acc

  defp addr_subnet({ip, prefix_len}) do
    %IP.Subnet{routing_prefix: ip, bit_length: prefix_len}
    |> IP.Subnet.to_string()
  end

  defp maybe_add_dns(ifname, %{dns: dns} = config) do
    {:fun, VintageNet.NameResolver, :setup, [ifname, config[:domain], dns]}
  end

  defp maybe_add_dns(_, _), do: []

  defp route_ip(ifname, addr) do
    {:run, "ip", ["route", "add", addr_subnet(addr), "dev", ifname]}
  end
end
